scale-decode-derive/src/lib.rs

```diff
@@ -0,0 +1,561 @@
+                val.decode_as_type().map_err(|e| e.at_field(#field_name))?
+            }),
+            quote!(#field_ident: {
+                let val = vals.next().unwrap()?;
+                val.decode_as_type().map_err(|e| e.at_field(#field_name))?
+            })
+        )
+    });
+
+    // if we skip any fields, we won't expect that field to exist in some tuple that's being given back.
+    let field_count = field_keyval_impls.clone().filter(|f| f.0).count();
+    let field_composite_keyvals = field_keyval_impls.clone().map(|v| v.1);
+    let field_tuple_keyvals = field_keyval_impls.map(|v| v.2);
+
+    (field_count, field_composite_keyvals, field_tuple_keyvals)
+}
+
+// Given some unnamed fields, generate impls like `get_field_value()` for each field. Do this for a tuple style impl.
+fn unnamed_field_vals<'f>(
+    _path_to_scale_decode: &'f syn::Path,
+    fields: &'f syn::FieldsUnnamed,
+) -> (usize, impl Iterator<Item = TokenStream2> + 'f) {
+    let field_val_impls = fields.unnamed.iter().enumerate().map(|(idx, f)| {
+        let field_attrs = FieldAttrs::from_attributes(&f.attrs).unwrap_or_default();
+        let skip_field = field_attrs.skip;
+
+        // If a field is skipped, we expect it to have a Default impl to use to populate it instead.
+        if skip_field {
+            return (false, quote!(::std::default::Default::default()));
+        }
+
+        (
+            true,
+            quote!({
+                let val = vals.next().unwrap()?;
+                val.decode_as_type().map_err(|e| e.at_idx(#idx))?
+            }),
+        )
+    });
+
+    // if we skip any fields, we won't expect that field to exist in some tuple that's being given back.
+    let field_count = field_val_impls.clone().filter(|f| f.0).count();
+    let field_vals = field_val_impls.map(|v| v.1);
+
+    (field_count, field_vals)
+}
+
+fn handle_generics<'a>(
+    attrs: &TopLevelAttrs,
+    generics: &'a syn::Generics,
+) -> (syn::ImplGenerics<'a>, syn::TypeGenerics<'a>, syn::WhereClause, syn::Type) {
+    let path_to_crate = &attrs.crate_path;
+    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
+
+    let mut where_clause = where_clause.cloned().unwrap_or(syn::parse_quote!(where));
+
+    if let Some(where_predicates) = &attrs.trait_bounds {
+        // if custom trait bounds are given, append those to the where clause.
+        where_clause.predicates.extend(where_predicates.clone());
+    } else {
+        // else, append our default bounds to each parameter to ensure that it all lines up with our generated impls and such:
+        for param in generics.type_params() {
+            let ty = &param.ident;
+            where_clause.predicates.push(syn::parse_quote!(#ty: #path_to_crate::IntoVisitor));
+            where_clause.predicates.push(syn::parse_quote!(#ty: #path_to_crate::IntoVisitor));
```

Comment:
nit: Are we pushing twice the same generic bound `T: IntoVisitor`? We could remove one of those.

Reply:
Ooh good spot; I guess this was a leftover from when I needed both lines for something different!